x/transferinject/ibc_middleware.go

```diff
@@ -0,0 +1,165 @@
+	clienttypes "github.com/cosmos/ibc-go/v6/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v6/modules/core/05-port/types"
+
+	"github.com/dymensionxyz/dymension/v3/x/transferinject/types"
+)
+
+type IBCSendMiddleware struct {
+	porttypes.ICS4Wrapper
+
+	rollappKeeper types.RollappKeeper
+	bankKeeper    types.BankKeeper
+}
+
+// NewIBCSendMiddleware creates a new ICS4Wrapper.
+// It intercepts outgoing IBC packets and adds token metadata to the memo if the rollapp doesn't have it.
+// This is a solution for adding token metadata to fungible tokens transferred over IBC,
+// targeted at rollapps that don't have the token metadata for the token being transferred.
+// More info here: https://www.notion.so/dymension/ADR-x-IBC-Denom-Metadata-Transfer-From-Hub-to-Rollapp-d3791f524ac849a9a3eb44d17968a30b
+func NewIBCSendMiddleware(
+	ics porttypes.ICS4Wrapper,
+	rollappKeeper types.RollappKeeper,
+	bankKeeper types.BankKeeper,
+) *IBCSendMiddleware {
+	return &IBCSendMiddleware{
+		ICS4Wrapper:   ics,
+		rollappKeeper: rollappKeeper,
+		bankKeeper:    bankKeeper,
+	}
+}
+
+// SendPacket wraps IBC ChannelKeeper's SendPacket function
+func (m *IBCSendMiddleware) SendPacket(
+	ctx sdk.Context,
+	chanCap *capabilitytypes.Capability,
+	destinationPort string, destinationChannel string,
+	timeoutHeight clienttypes.Height,
+	timeoutTimestamp uint64,
+	data []byte,
+) (sequence uint64, err error) {
+	packet := new(transfertypes.FungibleTokenPacketData)
+	if err = types.ModuleCdc.UnmarshalJSON(data, packet); err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrJSONUnmarshal, "unmarshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	if types.MemoAlreadyHasPacketMetadata(packet.Memo) {
+		return 0, types.ErrMemoTransferInjectAlreadyExists
+	}
+
+	rollapp, err := m.rollappKeeper.ExtractRollappFromChannel(ctx, destinationPort, destinationChannel)
+	if err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrInvalidRequest, "extract rollapp from packet: %s", err.Error())
+	}
+
+	// TODO: currently we check if receiving chain is a rollapp, consider that other chains also might want this feature
+	// meaning, find a better way to check if the receiving chain supports this middleware
+	if rollapp == nil {
+		return m.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+	}
+
+	// Check if the rollapp already contains the denom metadata by matching the base of the denom metadata.
+	// At the first match, we assume that the rollapp already contains the metadata.
+	// It would be technically possible to have a race condition where the denom metadata is added to the rollapp
+	// from another packet before this packet is acknowledged.
+	if Contains(rollapp.RegisteredDenoms, packet.Denom) {
+		return m.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
```

Comment:
here you use packet.Denom but in the ack code you use base
Can it be consistent?